Fix  #1497:Set up code coverage for Robolectric tests

peculiaruc opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
  Fixes #1497 Set up code coverage for Robolectric tests

### Output
![domains](https://user-images.githubusercontent.com/35475543/111076502-d07e0480-84ec-11eb-98af-2781139e682a.png)


![cli code coverage4](https://user-images.githubusercontent.com/35475543/110190606-672f2f00-7e24-11eb-9750-d0e53ba36c56.png)

<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

peculiaruc committed:
Code coverage for Roboelectric

peculiaruc committed:
Code coverage for Roboelectric

peculiaruc commented:
1.The document below contains my findings about running code coverages to determine the percentage of the app code covered by each unit test.
2. It also contain steps to configure the android studio to get code coverage reports.

https://docs.google.com/document/d/1oSIY-12lGl28MVgGxs0mbuJFTORkjGqypeiOGq95jLc/edit

It also await review from reviewers.

peculiaruc commented:
Github gist below  contains some earlier interactions on this issue. 

https://gist.github.com/peculiaruc/2e63953249c1af8d731246f737fec588

BenHenning commented on app/build.gradle:
```diff
@@ -191,9 +197,41 @@ dependencies {
   implementation project(":utility")
 }
+repositories {
+  mavenCentral()
```
This shouldn't be necessary since we define this in the root Gradle file. Can you try removing it to see if JaCoCo works?

peculiaruc replied:
Done

BenHenning commented on app/build.gradle:
```diff
@@ -191,9 +197,41 @@ dependencies {
+    csv.enabled false
+  }
+  def fileFilter = [
+      '**/R.class', '**/R$*.class', '**/BuildConfig.*', '**/Manifest*.*', '**/*Test*.*', 'android/**/*.*', '**/*$[0-9].*'
```
Why are these filters needed?

peculiaruc replied:
We don't actually need this for now

BenHenning commented on app/build.gradle:
```diff
@@ -191,9 +197,41 @@ dependencies {
+task jacocoTestReport(type: JacocoReport, dependsOn: ['testDebugUnitTest', 'createDebugCoverageReport']) {
+
+  reports {
+    xml.enabled = true
+    html.enabled = true
+    csv.enabled false
```
Do we need all 3?

peculiaruc replied:
No. just 2

BenHenning commented on build.gradle:
```diff
@@ -26,6 +29,16 @@ allprojects {
+  resolutionStrategy {
+    eachDependency { details ->
+      if (details.requested.group == 'org.jacoco') {
+        details.useVersion jacocoVersion
```
This seems a bit hacky--why is this needed?

peculiaruc replied:
This is not necessary too

BenHenning commented on build.gradle:
```diff
@@ -3,17 +3,20 @@
     google()
     jcenter()
     gradlePluginPortal()
+    mavenCentral()
```
I think we inadvertently add this in domain/build.gradle. Since you're adding this here, can you remove it from that Gradle file? It's a nice opportunistic fix.

peculiaruc replied:
Done

BenHenning commented on data/build.gradle:
```diff
@@ -2,6 +2,12 @@ apply plugin: 'com.android.library'
 apply plugin: 'kotlin-kapt'
+apply plugin: 'jacoco'
+
+jacoco {
+  toolVersion = "0.8.2"
+  version "0.7.1.201405082137"
+}
```
We should define this in one location to reduce potential error in the future if we ever change the version. It can be defined in one location similar to Kotlin version: https://github.com/oppia/oppia-android/blob/develop/build.gradle#L4.

Ditto elsewhere.

peculiaruc replied:
Done

BenHenning commented on domain/build.gradle:
```diff
@@ -12,6 +19,8 @@ android {
     targetSdkVersion 29
     versionCode 1
     versionName "1.0"
+    multiDexEnabled true
+    testInstrumentationRunner "org.oppia.android.testing.OppiaTestRunner"
```
We don't have instrumentation tests for domain--why is this needed?

peculiaruc replied:
It is actually not necessary

BenHenning commented on domain/build.gradle:
```diff
@@ -107,3 +124,30 @@ repositories {
+tasks.withType(Test) {
+  jacoco.includeNoLocationClasses = true
+}
+task jacocoTestReport(type: JacocoReport, dependsOn: ['testDebugUnitTest', 'createDebugCoverageReport']) {
```
Can we somehow share the JaCoCo task across all Gradle files that need it? Also, shouldn't this get included in app, domain, data, utility, & testing module Gradle files? I think you might be missing data.

peculiaruc replied:
For this, there is an error about setVersion.  See the gist below
https://gist.github.com/peculiaruc/54c08bd429a4fdf5f29ec3906f4f8db6

BenHenning reviewed (changes requested):
Thanks @peculiaruc! Some thoughts:

1. It seems like there are some potentially unnecessary changes in the PR. As general feedback: try to make sure your PRs only include exactly the changes you think are needed (& make sure you can justify why that is if prompted during review).
2. Try to minimize code duplication by reusing things like constants & tasks/methods where possible. I think there might be good potential for that in this PR.
3. Please make sure to assign the PR over to reviewers when it's ready for review or follow-up passes, otherwise it's likely to get missed. I suggest re-reading https://github.com/oppia/oppia-android/wiki/Guidance-on-submitting-a-PR#clarification-regarding-assignees-and-reviewers-section.

anandwana001 reviewed (changes requested):
From this PR, I tried running the task, but it get failed. 

```
./gradlew :domain:jacocoTestReport
```

```
> Task :utility:compileDebugJavaWithJavac FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':utility:compileDebugJavaWithJavac'.
> Compilation failed; see the compiler error output for details.

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.

* Get more help at https://help.gradle.org

Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
Use '--warning-mode all' to show the individual deprecation warnings.
See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings

BUILD FAILED in 8m 22s
64 actionable tasks: 64 executed
```

MaskedCarrot commented:
> 1.The document below contains my findings about running code coverages to determine the percentage of the app code covered by each unit test.
> 2. It also contain steps to configure the android studio to get code coverage reports.
> 
> https://docs.google.com/document/d/1oSIY-12lGl28MVgGxs0mbuJFTORkjGqypeiOGq95jLc/edit
> 
> It also await review from reviewers.

@peculiaruc can you please grant me access to this doc. My email is 2000apoorv@gmail.com. I have emailed you access request.
Thanks

BenHenning committed:
Merge branch 'develop' into setup_code_coverage_for_Roboelectric

MaskedCarrot commented:
It would be better if we can have JaCoCO in a separate file. I did this by creating a `jacoco.gradle` file in the project root [here](https://gist.github.com/MaskedCarrot/291390581791879e211edc8873eaf951) is my `jacoco.gradle` file. 

@peculiaruc PTAL.

peculiaruc committed:
updates code coverage

peculiaruc commented:
> From this PR, I tried running the task, but it get failed.
> 
> ```
> ./gradlew :domain:jacocoTestReport
> ```
> 
> ```
> > Task :utility:compileDebugJavaWithJavac FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > Compilation failed; see the compiler error output for details.
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> Use '--warning-mode all' to show the individual deprecation warnings.
> See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 8m 22s
> 64 actionable tasks: 64 executed
> ```

I have resolved this

peculiaruc committed:
Merge remote-tracking branch 'remotes/origin/setup_code_coverage_for_Roboelectric' into setup_code_coverage_for_Roboelectric

BenHenning commented:
Hi @peculiaruc. As mentioned in chat, I'm not actually sure what I'm supposed to be reviewing here since it seems there are still open conversation threads above, and this is in draft form. Can you please clarify what you'd like my thoughts on here?

Thanks!

peculiaruc commented:
> > 1.The document below contains my findings about running code coverages to determine the percentage of the app code covered by each unit test.
> > 2. It also contain steps to configure the android studio to get code coverage reports.
> > https://docs.google.com/document/d/1oSIY-12lGl28MVgGxs0mbuJFTORkjGqypeiOGq95jLc/edit
> > It also await review from reviewers.
> 
> @peculiaruc can you please grant me access to this doc. My email is [2000apoorv@gmail.com](mailto:2000apoorv@gmail.com). I have emailed you access request.
> Thanks

Seen

peculiaruc commented:
> It would be better if we can have JaCoCO in a separate file. I did this by creating a `jacoco.gradle` file in the project root [here](https://gist.github.com/MaskedCarrot/291390581791879e211edc8873eaf951) is my `jacoco.gradle` file.
> 
> @peculiaruc PTAL.

Seen

peculiaruc commented:
> Hi @peculiaruc. As mentioned in chat, I'm not actually sure what I'm supposed to be reviewing here since it seems there are still open conversation threads above, and this is in draft form. Can you please clarify what you'd like my thoughts on here?
> 
> Thanks!

I have resolved all the conversations. the pr is ready for review now

anandwana001 commented:
> > From this PR, I tried running the task, but it get failed.
> > ```
> > ./gradlew :domain:jacocoTestReport
> > ```
> > 
> > 
> > ```
> > > Task :utility:compileDebugJavaWithJavac FAILED
> > 
> > FAILURE: Build failed with an exception.
> > 
> > * What went wrong:
> > Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > > Compilation failed; see the compiler error output for details.
> > 
> > * Try:
> > Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> > 
> > * Get more help at https://help.gradle.org
> > 
> > Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> > Use '--warning-mode all' to show the individual deprecation warnings.
> > See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> > 
> > BUILD FAILED in 8m 22s
> > 64 actionable tasks: 64 executed
> > ```
> 
> I have resolved this

Getting the same error. Could you add a screenshot showing it is passing for you? @peculiaruc

anandwana001 commented:
@BenHenning Quick thought, we are trying adding `jacoco` here in Gradle tasks, but when we remove the Gradle everything goes off, should we divert this PR in a way to just have `jacoco` using Bazel, which does seem like a big task in terms of research too.  
Or should we carry forward with the Gradle stuff?

BenHenning commented:
We should carry on with Gradle. I think I mentioned somewhere else, but
Bazel will require actual changes to Bazel itself to support code coverage
(I dug into it enough to figure that out). It's not great adding another
dependency on Gradle, but I think it's more important to get code coverage
in place than finishing the Bazel migration.

On Wed, Feb 24, 2021 at 8:55 PM Akshay Nandwana <notifications@github.com>
wrote:

> @BenHenning <https://github.com/BenHenning> Quick thought, we are trying
> adding jacoco here in Gradle tasks, but when we remove the Gradle
> everything goes off, should we divert this PR in a way to just have jacoco
> using Bazel, which does seem like a big task in terms of research too.
> Or should we carry forward with the Gradle stuff?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/2730#issuecomment-785611252>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADDB3PTTCOJBMEJV2BEYXD3TAXJ2JANCNFSM4XWVYMIQ>
> .
>

peculiaruc commented:
> > > From this PR, I tried running the task, but it get failed.
> > > ```
> > > ./gradlew :domain:jacocoTestReport
> > > ```
> > > 
> > > 
> > > ```
> > > > Task :utility:compileDebugJavaWithJavac FAILED
> > > 
> > > FAILURE: Build failed with an exception.
> > > 
> > > * What went wrong:
> > > Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > > > Compilation failed; see the compiler error output for details.
> > > 
> > > * Try:
> > > Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> > > 
> > > * Get more help at https://help.gradle.org
> > > 
> > > Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> > > Use '--warning-mode all' to show the individual deprecation warnings.
> > > See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> > > 
> > > BUILD FAILED in 8m 22s
> > > 64 actionable tasks: 64 executed
> > > ```
> > 
> > 
> > I have resolved this
> 
> Getting the same error. Could you add a screenshot showing it is passing for you? @peculiaruc

I didn't take a screenshot , Trying to get this again was a bit headache with windows. My thought about using CLI, since we can easily get our code coverage report with Andoid studio, i think we can go with that option. What do you think please ?

anandwana001 commented:
The reason we cannot opt for Android Studio option is because we need to automate this task and want to run in the GitHub Actions.

peculiaruc commented:
> The reason we cannot opt for Android Studio option is because we need to automate this task and want to run in the GitHub Actions.

Ok. I am on it. @MaskedCarrot  do you have any suggestion about this?

MaskedCarrot commented:
@peculiaruc this issue is caused because we are using `noLocationCalsses` changing `tasks.withType(Test)` as mentioned below fixes the problem.
```
tasks.withType(Test) {
  jacoco.includeNoLocationClasses = true
  jacoco.excludes = ['jdk.internal.*']
}
```

BenHenning commented:
Adding to @anandwana001's point: getting this to work in Android Studio is a nice-to-have, but on CLI is a must. Also, if we can't get it working in Android Studio then that's probably completely fine since people can always run the CLI version. Plus, in the Bazel world we probably won't have Android Studio support, anyway.

@peculiaruc I'm going to de-assign myself for now since it seems like there's still work needed here to get code coverage working as expected. Please add a screenshot to the PR's opening comment showing code coverage results once this is working, and reassign to me at that point.

peculiaruc commented:
Noted

On Fri, 26 Feb 2021, 4:58 AM Ben Henning, <notifications@github.com> wrote:

> Adding to @anandwana001 <https://github.com/anandwana001>'s point:
> getting this to work in Android Studio is a nice-to-have, but on CLI is a
> must. Also, if we can't get it working in Android Studio then that's
> probably completely fine since people can always run the CLI version. Plus,
> in the Bazel world we probably won't have Android Studio support, anyway.
>
> @peculiaruc <https://github.com/peculiaruc> I'm going to de-assign myself
> for now since it seems like there's still work needed here to get code
> coverage working as expected. Please add a screenshot to the PR's opening
> comment showing code coverage results once this is working, and reassign to
> me at that point.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/2730#issuecomment-786394622>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIOVAV4O5OLTHO6MZH72QA3TA4L6VANCNFSM4XWVYMIQ>
> .
>

MaskedCarrot commented on app/build.gradle:
```diff
@@ -191,9 +192,29 @@ dependencies {
+
+  sourceDirectories.setFrom(files([mainSrc]))
+  executionData.setFrom(fileTree(dir: project.buildDir, includes: [
+      'jacoco/testDebugUnitTest.exec', 'outputs/code_coverage/debugAndroidTest/connected/**/*.ec'
```
I think  `outputs/code_coverage/debugAndroidTest/connected/**/*.ec`  is not needed here as we only want to cover robolectric tests. Removing this makes the Gradle task work fine.

MaskedCarrot replied:
Also if we keep this i think we will have to downgrade the version of Gradle because Gradle has an ongoing issue which mentions that since version 6.4  Gradle does not wait for the coverage report to be generated.

MaskedCarrot replied:
[here](https://github.com/gradle/gradle/issues/14132) is the link to gradle's issue.

peculiaruc replied:
Seen

MaskedCarrot reviewed (commented):
@peculiaruc PTAL at the following comments

MaskedCarrot commented:
> @peculiaruc this issue is caused because we are using `noLocationCalsses` changing `tasks.withType(Test)` as mentioned below fixes the problem.
> 
> ```
> tasks.withType(Test) {
>   jacoco.includeNoLocationClasses = true
>   jacoco.excludes = ['jdk.internal.*']
> }
> ```
**update** 
 doing this does not fix the issue we are facing with `./gradlew :domain:jacocoTestReport` but we should add this for support for java9+ as mentioed over [here](https://github.com/gradle/gradle/issues/5184#issuecomment-457865951).
WDYT @peculiaruc @anandwana001?

peculiaruc commented:
> > @peculiaruc this issue is caused because we are using `noLocationCalsses` changing `tasks.withType(Test)` as mentioned below fixes the problem.
> > ```
> > tasks.withType(Test) {
> >   jacoco.includeNoLocationClasses = true
> >   jacoco.excludes = ['jdk.internal.*']
> > }
> > ```
> 
> **update**
> doing this does not fix the issue we are facing with `./gradlew :domain:jacocoTestReport` but we should add this for support for java9+ as mentioed over [here](https://github.com/gradle/gradle/issues/5184#issuecomment-457865951).
> WDYT @peculiaruc @anandwana001?

Yes It doesn't fix it as i mention in the chat but we can keep it

peculiaruc committed:
Updated gradle to work with CLI

peculiaruc commented:
![cli code coverage1](https://user-images.githubusercontent.com/35475543/109733276-11a61880-7bbf-11eb-91a7-69b52af674df.png)
![cli code coverage2](https://user-images.githubusercontent.com/35475543/109733291-1a96ea00-7bbf-11eb-9dd0-007653b1f036.png)

peculiaruc commented:
> > > From this PR, I tried running the task, but it get failed.
> > > ```
> > > ./gradlew :domain:jacocoTestReport
> > > ```
> > > 
> > > 
> > > ```
> > > > Task :utility:compileDebugJavaWithJavac FAILED
> > > 
> > > FAILURE: Build failed with an exception.
> > > 
> > > * What went wrong:
> > > Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > > > Compilation failed; see the compiler error output for details.
> > > 
> > > * Try:
> > > Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> > > 
> > > * Get more help at https://help.gradle.org
> > > 
> > > Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> > > Use '--warning-mode all' to show the individual deprecation warnings.
> > > See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> > > 
> > > BUILD FAILED in 8m 22s
> > > 64 actionable tasks: 64 executed
> > > ```
> > 
> > 
> > I have resolved this
> 
> Getting the same error. Could you add a screenshot showing it is passing for you? @peculiaruc

@anandwana001 PTAL

peculiaruc commented:
![cli code coverage1](https://user-images.githubusercontent.com/35475543/109736512-bbd46f00-7bc4-11eb-8e4e-c7f996009378.png)
![cli code coverage3](https://user-images.githubusercontent.com/35475543/109736523-c0992300-7bc4-11eb-8b3b-4c6f9ddf37eb.png)

BenHenning commented:
@peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.

Please reassign once:
1. The opening comment explains the PR & shows code coverage working
2. You have another message in the comment thread asking for either a specific check or for full-review of the PR

I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.

Thanks!

MaskedCarrot commented:
@peculiaruc I tried  `./gradlew domain:jacocoTestReport`, it failed with the exception 
 ```
* What went wrong:
Execution failed for task ':domain:createDebugAndroidTestCoverageReport'.
```

This is because as mentioned in my earlier [comment](https://github.com/oppia/oppia-android/pull/2730#discussion_r584324269) gradle does not wait for the report to be generated so we will have to downgrade the version of Gradle. 
Please correct me if I am wrong, since we have removed  `outputs/code_coverage/debugAndroidTest/connected/**/*.ec` we don't need `createDebugCoverageReport` anyway.

MaskedCarrot commented on app/build.gradle:
```diff
@@ -191,9 +192,29 @@ dependencies {
+    xml.enabled = true
+    html.enabled = true
+  }
+  def mainSrc = "$project.projectDir/src/main/kotlin"
```
I think the path here should be `$project.projectDir/src/main/java`.

peculiaruc replied:
Yes i think so

MaskedCarrot commented on data/build.gradle:
```diff
@@ -116,10 +116,10 @@ configurations {
 
-tasks.withType(Test) {
+ tasks.withType(Test) {
   jacoco.includeNoLocationClasses = true
```
we will have to add `jacoco.excludes = ['jdk.internal.*']` in all the build.gradle files.

peculiaruc replied:
This won't make a difference now but i will still add it up

peculiaruc commented:
> @peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.
> 
> Please reassign once:
> 
> 1. The opening comment explains the PR & shows code coverage working
> 2. You have another message in the comment thread asking for either a specific check or for full-review of the PR
> 
> I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.
> 
> Thanks!

Noted @BenHenning .

> From this PR, I tried running the task, but it get failed.
> 
> ```
> ./gradlew :domain:jacocoTestReport
> ```
> 
> ```
> > Task :utility:compileDebugJavaWithJavac FAILED
> 
> FAILURE: Build failed with an exception.
> 
> * What went wrong:
> Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > Compilation failed; see the compiler error output for details.
> 
> * Try:
> Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> 
> * Get more help at https://help.gradle.org
> 
> Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> Use '--warning-mode all' to show the individual deprecation warnings.
> See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> 
> BUILD FAILED in 8m 22s
> 64 actionable tasks: 64 executed
> ```

@BenHenning the last push i made Fixes part of this PR . Specifically fix this ./gradlew :domain:jacocoTestReport

peculiaruc commented:
> > @peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.
> > Please reassign once:
> > 
> > 1. The opening comment explains the PR & shows code coverage working
> > 2. You have another message in the comment thread asking for either a specific check or for full-review of the PR
> > 
> > I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.
> > Thanks!
> 
> Noted @BenHenning .
> 
> > From this PR, I tried running the task, but it get failed.
> > ```
> > ./gradlew :domain:jacocoTestReport
> > ```
> > 
> > 
> > ```
> > > Task :utility:compileDebugJavaWithJavac FAILED
> > 
> > FAILURE: Build failed with an exception.
> > 
> > * What went wrong:
> > Execution failed for task ':utility:compileDebugJavaWithJavac'.
> > > Compilation failed; see the compiler error output for details.
> > 
> > * Try:
> > Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights.
> > 
> > * Get more help at https://help.gradle.org
> > 
> > Deprecated Gradle features were used in this build, making it incompatible with Gradle 7.0.
> > Use '--warning-mode all' to show the individual deprecation warnings.
> > See https://docs.gradle.org/6.8/userguide/command_line_interface.html#sec:command_line_warnings
> > 
> > BUILD FAILED in 8m 22s
> > 64 actionable tasks: 64 executed
> > ```
> 
> @BenHenning the last push i mad Fixes part of this PR . Specifically fix this ./gradlew :domain:jacocoTestReport

peculiaruc commented:
> @peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.
> 
> Please reassign once:
> 
> 1. The opening comment explains the PR & shows code coverage working
> 2. You have another message in the comment thread asking for either a specific check or for full-review of the PR
> 
> I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.
> 
> Thanks!

@BenHenning  the last push i made Fixes part of this PR . Specifically fix this ./gradlew :domain:jacocoTestReport. I modified the domain module, then ran code coverage with CLI.
this screenshot https://github.com/oppia/oppia-android/pull/2730#issuecomment-789344187 is the result of test passed running ./gadlew :domain:jacocoTestReport.

peculiaruc commented:
I also reorganized the doc so that headers are consistent in casing & wording and included the recent modification in the gradle

peculiaruc commented:
> @peculiaruc I tried `./gradlew domain:jacocoTestReport`, it failed with the exception
> 
> ```
> * What went wrong:
> Execution failed for task ':domain:createDebugAndroidTestCoverageReport'.
> ```
> 
> This is because as mentioned in my earlier [comment](https://github.com/oppia/oppia-android/pull/2730#discussion_r584324269) gradle does not wait for the report to be generated so we will have to downgrade the version of Gradle.
> Please correct me if I am wrong, since we have removed `outputs/code_coverage/debugAndroidTest/connected/**/*.ec` we don't need `createDebugCoverageReport` anyway.

This passed locally for me, have a look https://github.com/oppia/oppia-android/pull/2730#issuecomment-789344187

MaskedCarrot commented:
@peculiaruc  can you add a screenshot of the jacoco report generated for domain module

peculiaruc committed:
Updated all module to work with CLI

peculiaruc commented:
> @peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.
> 
> Please reassign once:
> 
> 1. The opening comment explains the PR & shows code coverage working
> 2. You have another message in the comment thread asking for either a specific check or for full-review of the PR
> 
> I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.
> 
> Thanks!
 F or 1 and 2 done

The whole PR is ready for review.  In this last commit , i modified all the gradle to run code coverage with ./gradlew :domain:jacocoTestReport in all the module.
I also changed kotlin to java def mainSrc = "$project.projectDir/src/main/kotlin"
I also modified the doc to contain the recent modification.

Screenshot for code passed in the terminal:

![cli code coverage1](https://user-images.githubusercontent.com/35475543/109802164-c40dc800-7c1f-11eb-808c-dd951f450d3a.png)

peculiaruc commented:
Screenshot for the test result
![cli code coverage4](https://user-images.githubusercontent.com/35475543/109802957-bd338500-7c20-11eb-84b9-cb5759faef3a.png)

peculiaruc commented:
> @peculiaruc can you add a screenshot of the jacoco report generated for domain module

Done

BenHenning commented:
@peculiaruc please also add the requested screenshots from https://github.com/oppia/oppia-android/pull/2730#issuecomment-789403671 to https://github.com/oppia/oppia-android/pull/2730#issue-574290460.

Also, is the error mentioned in https://github.com/oppia/oppia-android/pull/2730#issuecomment-789506631 still occurring for you?

@MaskedCarrot you mentioned in the meeting earlier that there are some runtime failures with JaCoCo 0.8.6 for the domain module. Are those observable in this PR? I don't see coverage CI workflows running. Am I missing them?

@MaskedCarrot @peculiaruc are we going to consolidate #2744 & this PR? There seem to be differences, and I want to make sure we're requesting submission of a single, cohesive solution before going through a full code review. Significant late changes make reviewing code harder. :)

peculiaruc commented:
> @peculiaruc please also add the requested screenshots from [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issuecomment-789403671) to [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issue-574290460).
> 
> Also, is the error mentioned in [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issuecomment-789506631) still occurring for you?
> 
> @MaskedCarrot you mentioned in the meeting earlier that there are some runtime failures with JaCoCo 0.8.6 for the domain module. Are those observable in this PR? I don't see coverage CI workflows running. Am I missing them?
> 
> @MaskedCarrot @peculiaruc are we going to consolidate #2744 & this PR? There seem to be differences, and I want to make sure we're requesting submission of a single, cohesive solution before going through a full code review. Significant late changes make reviewing code harder. :)

@BenHenning I will add the screenshots . The error is occurring again.
For #2744 I have asked @MaskedCarrot  to pull it out of this pr so that the code review will be easy. Other help from him should be on this main pr I think

peculiaruc commented:
> @peculiaruc when assigning PRs please include a message explaining specifically what changed, and what we should be looking for next. I'm not actually sure from your previous message. Should we be reviewing the entirety of the PR? Also per my earlier comment, you should be adding screenshots specifically of a working code coverage result run (e.g. the actual results & the tests passing) in the [opening comment](https://github.com/oppia/oppia-android/pull/2730#issue-574290460) of the PR thread. That comment should have a general explanation for what the PR is doing, and demonstrate how it's working. See other PRs for an example of how this is sometimes done.
> 
> Please reassign once:
> 
> 1. The opening comment explains the PR & shows code coverage working
> 2. You have another message in the comment thread asking for either a specific check or for full-review of the PR
> 
> I'm happy to look at the whole PR/anything specific you ask re: (2), I just to know what needs to be reviewed before I can take a look.
> 
> Thanks!

Done

peculiaruc commented:
> @peculiaruc please also add the requested screenshots from [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issuecomment-789403671) to [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issue-574290460).
> 
> Also, is the error mentioned in [#2730 (comment)](https://github.com/oppia/oppia-android/pull/2730#issuecomment-789506631) still occurring for you?
> 
> @MaskedCarrot you mentioned in the meeting earlier that there are some runtime failures with JaCoCo 0.8.6 for the domain module. Are those observable in this PR? I don't see coverage CI workflows running. Am I missing them?
> 
> @MaskedCarrot @peculiaruc are we going to consolidate #2744 & this PR? There seem to be differences, and I want to make sure we're requesting submission of a single, cohesive solution before going through a full code review. Significant late changes make reviewing code harder. :)

Done

BenHenning commented:
Thanks for adding the screenshot!

To be clear per my earlier message: I don't want to lose any work done here. Is it possible for us to consolidate to a single solution for setting up the JaCoCo configuration (e.g. #2744 has a nice approach where it just sets up the configuration in 1 place and replicates it for each module)? The GitHub actions bits can wait until later (probably the GSoC project for actually introducing code coverage), but we should aim to get a working JaCoCo configuration for all of the app's modules as part of this PR.

@peculiaruc can you update this PR to match the configuration setup done in #2744?

@peculiaruc / @MaskedCarrot can you verify whether this PR, after being updated, successfully runs code coverage for each of: app, data, testing, domain, utility?

peculiaruc commented:
> Thanks for adding the screenshot!
> 
> To be clear per my earlier message: I don't want to lose any work done here. Is it possible for us to consolidate to a single solution for setting up the JaCoCo configuration (e.g. #2744 has a nice approach where it just sets up the configuration in 1 place and replicates it for each module)? The GitHub actions bits can wait until later (probably the GSoC project for actually introducing code coverage), but we should aim to get a working JaCoCo configuration for all of the app's modules as part of this PR.
> 
> @peculiaruc can you update this PR to match the configuration setup done in #2744?
> 
> @peculiaruc / @MaskedCarrot can you verify whether this PR, after being updated, successfully runs code coverage for each of: app, data, testing, domain, utility?

Oh! I understand now. Sure. @MaskedCarrot please confirm @BenHenning request from your end. Thanks

MaskedCarrot commented:
@peculiaruc can you please update this pr do that I can test it locally

peculiaruc committed:
Created Jacoco.gradle

peculiaruc committed:
Created Jacoco.gradle

peculiaruc commented:
![jacoco report](https://user-images.githubusercontent.com/35475543/110216800-5ed10580-7eb1-11eb-888b-c034ed3f1ba2.png)
![jacoco sumarry](https://user-images.githubusercontent.com/35475543/110216803-6395b980-7eb1-11eb-98be-19887f99cfc0.png)

peculiaruc commented:
> @peculiaruc can you please update this pr do that I can test it locally

@MaskedCarrot Please you can test this PR locally now

MaskedCarrot commented on app/build.gradle:
```diff
@@ -4,6 +4,7 @@ apply plugin: 'com.google.gms.google-services'
 apply plugin: 'kotlin-android'
 apply plugin: 'kotlin-android-extensions'
 apply plugin: 'kotlin-kapt'
+apply from: "$project.rootDir/jacoco.gradle"
```
please enable test coverage for app module

MaskedCarrot commented:
@peculiaruc I had a few suggestions please take a look

MaskedCarrot commented on jacoco.gradle:
```diff
@@ -0,0 +1,100 @@
+      "jacoco/testDebugUnitTest.exec"
+  ]))
+}
+  tasks.withType(Test) {
+  jacoco.includeNoLocationClasses = true
+}
```
I think we don't need this here. because we have already have this in  both the functions(`setupKotlinReporting` and `setupAndroidReporting`)

MaskedCarrot commented on model/build.gradle:
```diff
@@ -1,5 +1,6 @@
 apply plugin: 'java-library'
 apply plugin: 'com.google.protobuf'
+apply plugin: 'jacoco'
```
we don't have any tests in the model module so we can remove jacoco from `model/build.gradle`.

peculiaruc replied:
Yes taught i remove this , done

MaskedCarrot commented on testing/build.gradle:
```diff
@@ -24,6 +26,9 @@ android {
   }
 
```
test coverage is not enabled in the testing module so we should add this here.
```
  buildTypes {
    debug {
      testCoverageEnabled true
    }
  }
```

peculiaruc replied:
test coverage is enabled here.  I got the correct report here

MaskedCarrot commented on app/build.gradle:
```diff
@@ -4,6 +4,7 @@ apply plugin: 'com.google.gms.google-services'
 apply plugin: 'kotlin-android'
 apply plugin: 'kotlin-android-extensions'
 apply plugin: 'kotlin-kapt'
+apply from: "$project.rootDir/jacoco.gradle"
```
please enable test coverage for the app module.

peculiaruc replied:
Test coverage is enabled here. Please take a look at [this](https://github.com/oppia/oppia-android/pull/2730#issuecomment-792005962)

MaskedCarrot commented:
also one more suggestion, it will be nice to have  `group "Reporting"` in `setupAndroidReporting()` like we do in `setupKotlinReporting()`

MaskedCarrot commented:
This PR uses jacoco version 0.8.5. when i executed the coverage task I had the following findings

`for version 0.8.5`
1. correct coverage was reported for app and testing module.
2. 0% coverage was reported for domain,data and utility module.

`on changing the jacoco version to 0.7.9`
1. correct coverage was reported for domain, testing, data and utility module.
2. As for the app module the gradle task failed because most of the tests. were failing. The failing tests had the exception `java.lang.VerifyError: Stack map does not match the one at exception handler 359`

@peculiaruc @BenHenning  PTAL

peculiaruc commented:
> also one more suggestion, it will be nice to have `group "Reporting"` in `setupAndroidReporting()` like we do in `setupKotlinReporting()`

I think since we have group reporting in 'setuoKotlinReporting()', we don't need it again in `setupAndroidReporting()`. Adding it will cause jacocoTestReport to be missing.

MaskedCarrot commented:
> I think since we have group reporting in 'setuoKotlinReporting()', we don't need it again in `setupAndroidReporting()`. Adding it will cause jacocoTestReport to be missing.



@peculiaruc Please correct me if I am wrong here, 
I think  `group` is used to group similar tasks together so adding it will not cause jacocoTaskReport to be missing.

peculiaruc commented:
> This PR uses jacoco version 0.8.5. when i executed the coverage task I had the following findings
> 
> `for version 0.8.5`
> 
> 1. correct coverage was reported for app and testing module.
> 2. 0% coverage was reported for domain,data and utility module.
> 
> `on changing the jacoco version to 0.7.9`
> 
> 1. correct coverage was reported for domain, testing, data and utility module.
> 2. As for the app module the gradle task failed because most of the tests. were failing. The failing tests had the exception `java.lang.VerifyError: Stack map does not match the one at exception handler 359`
> 
> @peculiaruc @BenHenning PTAL

I used both of them as you can see in earlier push . I got the correct reports now. check the screenshot above. Nevertheless @BenHenning should also look into that

peculiaruc commented:
> > I think since we have group reporting in 'setuoKotlinReporting()', we don't need it again in `setupAndroidReporting()`. Adding it will cause jacocoTestReport to be missing.
> 
> @peculiaruc Please correct me if I am wrong here,
> I think `group` is used to group similar tasks together so adding it will not cause jacocoTaskReport to be missing.

I know but it kept throwing that error, that's why I didn't use that.

MaskedCarrot commented:
@peculiaruc  can you please check this again. It showed no error for me

peculiaruc commented:
> @peculiaruc can you please check this again. It showed no error for me

Please without the group Reporting. do you still get the correct test coverage?

MaskedCarrot commented:
yes, I do get coverage with or without `group`. 

But by adding `group "Reporting"` we can make the code more consistent.

peculiaruc commented:
> yes, I do get coverage with or without `group`.
> 
> But by adding `group "Reporting"` we can make the code more consistent.

defer to @BenHenning input

peculiaruc committed:
updated code coverage setup

MaskedCarrot commented:
> This PR uses jacoco version 0.8.5. when i executed the coverage task I had the following findings
> 
> `for version 0.8.5`
> 
> 1. correct coverage was reported for app and testing module.
> 2. 0% coverage was reported for domain,data and utility module.
> 
> `on changing the jacoco version to 0.7.9`
> 
> 1. correct coverage was reported for domain, testing, data and utility module.
> 2. As for the app module the gradle task failed because most of the tests. were failing. The failing tests had the exception `java.lang.VerifyError: Stack map does not match the one at exception handler 359`


For @BenHenning 's reference. 
[here](https://gist.github.com/MaskedCarrot/46f4a589a4c66f96c3bf5539229df1aa) is the link to the error in detail for a test in app module when using `jacoco version 0.7.9`

peculiaruc commented:
> > This PR uses jacoco version 0.8.5. when i executed the coverage task I had the following findings
> > `for version 0.8.5`
> > 
> > 1. correct coverage was reported for app and testing module.
> > 2. 0% coverage was reported for domain,data and utility module.
> > 
> > `on changing the jacoco version to 0.7.9`
> > 
> > 1. correct coverage was reported for domain, testing, data and utility module.
> > 2. As for the app module the gradle task failed because most of the tests. were failing. The failing tests had the exception `java.lang.VerifyError: Stack map does not match the one at exception handler 359`
> 
> For @BenHenning 's reference.
> [here](https://gist.github.com/MaskedCarrot/46f4a589a4c66f96c3bf5539229df1aa) is the link to the error in detail for a test in app module when using `jacoco version 0.7.9`

@MaskedCarrot what do you think can be solution to this error?

BenHenning commented:
@MaskedCarrot @peculiaruc can you please reply with the exact repro instructions for me? E.g. I'm looking for something like:

1. Check out branch XX
2. Open Android Studio
3. Run this command/click this button
4. Observe this output

Both for the working & failing modules. If there's anything else about my environment I need, please let me know. With repro instructions I can better understand what's going on, but without them it will be difficult to investigate.

MaskedCarrot commented:
@BenHenning To reproduce this
1. checkout branch `peculiaruc:setup_code_coverage_for_Roboelectric`
2. open android studio
3. run the gradle task `./gradlew domain:jacocoTestReport`
4. Find the coverage report in `domain/build/reports/jacoco/jacocoTestReport/html/index.html`
5. see  0% coverage reported for domain module.
6. (Note the version of jacoco just used was` 0.8.5`)
7. similarly 0 coverage is also reported for utility and data module on running `./gradlew utility:jacocoTestReport` and `./gradlew data:jacocoTestReport`
8. Correct Coverage is reported for testing and app moudle.
9. Now open `jacoco.gradle` file.
10. change the line 4  from `toolVersion = "0.8.5"` to `toolVersion = "0.7.9"`
11. run `./gradlew domain:jacocoTestReport` again
12. see correct coverage reported for domain, utility , testing and data module.
13. run `./gradlew app:jacocoTestReport` , build fails see the error.

BenHenning commented:
Hmm, domain tests seem to be failing for me locally. Particularly: ExceptionsControllerTest. Is that an issue for either of you @peculiaruc or @MaskedCarrot?

MaskedCarrot commented:
@BenHenning I executed these tests on manjaro and GitHub actions but I can't reproduce any error with domain module's task. All tests of ExceptionsControllerTest are passing for me.

BenHenning commented:
@peculiaruc can you please bring this up to date with the latest develop changes?

BenHenning commented:
Okay, I've been digging into this off and on for the past few hours and haven't yet made progress. I've come across a lot of cases of people reporting 0% coverage, and from what I can tell the issue starts breaking in JaCoCo 0.8.0. While 0.7.9 does provide code coverage for the utility module, it doesn't for app as was mentioned above (FWIW I don't see an error though, but I was using https://github.com/vanniktech/gradle-android-junit-jacoco-plugin to see if any of the advice from https://stackoverflow.com/a/33064500 would help--it didn't unfortunately). Interestingly, 0.7.8 also doesn't generate code coverage for utility, but that might be a quirk of the plugin above.

My best current guess is this has something to do with https://github.com/jacoco/jacoco/issues/996 which is currently unresolved, and the workarounds listed both there and in a related issue (https://issuetracker.google.com/issues/178015739) do not seem to actually help in these cases. It does seem like AGP and JaCoCo have a rather complex integration, and that's making this *much* harder to understand why one configuration works vs. another (and unfortunately the JaCoCo team doesn't seem particularly responsive).

I think we should still dig into this for the rest of the week, but if we can't find a workable solution we may need to cancel the code coverage GSoC project (FYI @rt4914 @vinitamurthi @Sarthak2601 @aggarwalpulkit596 @anandwana001).

peculiaruc commented:
> @BenHenning To reproduce this
> 
> 1. checkout branch `peculiaruc:setup_code_coverage_for_Roboelectric`
> 2. open android studio
> 3. run the gradle task `./gradlew domain:jacocoTestReport`
> 4. Find the coverage report in `domain/build/reports/jacoco/jacocoTestReport/html/index.html`
> 5. see  0% coverage reported for domain module.
> 6. (Note the version of jacoco just used was` 0.8.5`)
> 7. similarly 0 coverage is also reported for utility and data module on running `./gradlew utility:jacocoTestReport` and `./gradlew data:jacocoTestReport`
> 8. Correct Coverage is reported for testing and app moudle.
> 9. Now open `jacoco.gradle` file.
> 10. change the line 4  from `toolVersion = "0.8.5"` to `toolVersion = "0.7.9"`
> 11. run `./gradlew domain:jacocoTestReport` again
> 12. see correct coverage reported for domain, utility , testing and data module.
> 13. run `./gradlew app:jacocoTestReport` , build fails see the error.

Good job @MaskedCarrot

peculiaruc committed:
updated with develop branch

peculiaruc committed:
updated with develop branch

BenHenning commented:
@MaskedCarrot @peculiaruc I suggest continuing to try to make progress on this. I will try another avenue after I get the alpha MR2 in place (hopefully by this weekend).

peculiaruc commented:
> @MaskedCarrot @peculiaruc I suggest continuing to try to make progress on this. I will try another avenue after I get the alpha MR2 in place (hopefully by this weekend).

Noted

anandwana001 commented:
Closing this PR as no further update.

peculiaruc commented:
Still on it.

On Mon, 19 Apr 2021, 4:25 AM Akshay Nandwana, ***@***.***>
wrote:

> Closing this PR as no further update.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/oppia/oppia-android/pull/2730#issuecomment-822140057>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AIOVAV2QXPHI3DB54YNESGDTJOPBNANCNFSM4XWVYMIQ>
> .
>

anandwana001 commented:
@peculiaruc are you still working on this?

peculiaruc commented:
> @peculiaruc are you still working on this?

 Please, I am not in the right state of mind to work on it for now. sorry for the delay. Thanks

BenHenning commented:
Closing this for now, mainly due to the fact that we prefer using Bazel & changes will be needed in Bazel itself to unblock code coverage. This is a nice point of reference for when we conduct that work.
